### PR TITLE
[R4R]txindex/kv: Fsync data to disk immediately after receiving it 

### PR DESCRIPTION
**Pull request by guagualvcha**
Why this pr:
When restarting chain node, sometimes we lost tx index about recent(around 80)blocks, and some client complains that they can't find the tx by RPC call(tx_search) when the tx do exist in the block.

I try to partially fix this issue in a simple way by writing the index data in a sync way. 
There is no performance difference under 1K TPS according to our test.

It is still possible that lost index data after restarting the node, but only 2 block data will lost at most.
I try to totally fix this in https://github.com/tendermint/tendermint/pull/3847/files, but this one is simple and can solve most part of the issue. Please review first, thks.

### CODE REVIEW

**Comment by guagualvcha** [comment]
image

CI need fix

File: CHANGELOG_PENDING.md [hunk 1]
```diff
@@ -30,3 +30,4 @@ program](https://hackerone.com/tendermint).
 ### BUG FIXES:
 
 - [tools] [\#4023](https://github.com/tendermint/tendermint/issues/4023) Refresh `tm-monitor` health when validator count is updated (@erikgrinaker)
+- [index] [\#4104](https://github.com/tendermint/tendermint/pull/4104) Fix tx index lag too much
```
Comment (melekes) [inline CHANGELOG_PENDING.md]:
```suggestion
- [state] [\#4104](https://github.com/tendermint/tendermint/pull/4104) txindex/kv: Fsync data to disk immediately after receiving (@guagualvcha)
```

**Comment by melekes** [comment]
BEFORE:

```
BenchmarkTxIndex1-2               100000             12434 ns/op
BenchmarkTxIndex500-2                300           5151564 ns/op
BenchmarkTxIndex1000-2               100          15053910 ns/op
BenchmarkTxIndex2000-2               100          18238892 ns/op
BenchmarkTxIndex10000-2               20         124287930 ns/op
```

AFTER:

```
BenchmarkTxIndex1-2                 2000            795431 ns/op
BenchmarkTxIndex500-2                200           6385124 ns/op
BenchmarkTxIndex1000-2               100          11388219 ns/op
BenchmarkTxIndex2000-2               100          20514873 ns/op
BenchmarkTxIndex10000-2               20         107456004 ns/op
```

Performance drop is pretty steep, but I think it's the right thing to do UNTIL we have a WAL.

**Comment by ebuchman** [comment]
@melekes what is that benchmark on ? Individual `Index` calls or `AddBatch` calls? The former we don't seem to actually use other than in tests, the later I would expect much lesser performance hit from.

Regardless this is an important fix - thanks @guagualvcha . Seems also independent from https://github.com/tendermint/tendermint/issues/3211 which we'll need to address as well.

**Comment by melekes** [comment]
> @melekes what is that benchmark on ? Individual Index calls or AddBatch calls?

AddBatch calls

https://github.com/tendermint/tendermint/blob/97222eb90fc4b2882fe1859a665f2c84764c4b58/state/txindex/kv/kv_test.go#L314-L353

**Comment by ebuchman** [comment]
Hmm. Is the first number in `BenchmarkTxIndex2000-2` the number of txs? The cost is pretty high for the smaller nums but seems it's amortized for larger ones so maybe it's ok. Crazy that the cost is like 70x for the first one and otherwise only around ~1.2x or so for the others

**Comment by melekes** [comment]
> Is the first number in BenchmarkTxIndex2000-2 the number of txs?

Yes

**Pull request merged**